### ISSUE TITLE
Occasional SEGV

### ISSUE THREAD

**Issue by moewew** [plk/biber#87]
Apparently the `.bib` entry

```
@book{hak,
  author = {al- Hakim, Tawfik},
  title = {Title},
  date = {2015},
}
```

lets Biber fail occasionally with SEGV.

Here is an MWE

```latex
\documentclass{article}
\usepackage[ngerman]{babel}
\usepackage[babel,german=guillemets]{csquotes} 
\usepackage[backend=biber, style=authoryear]{biblatex}

\usepackage{filecontents}
\begin{filecontents*}{\jobname.bib}
@book{hak,
  author = {al- Hakim, Tawfik},
  title = {Title},
  date = {2015},
}
\end{filecontents*}

\addbibresource{\jobname.bib}


\usepackage{fontspec}
\setmainfont{Linux Libertine}

\begin{document}
\cite{hak}
\printbibliography
\end{document}
```

I can, however, not get the MWE to fail consistently, the Biber run will only fail every fourth or fifth time with a SEGV, sometimes Biber just hangs and doesn't finish, and more often that not the compilation works just fine.

I checked with Biber 2.2 from sourceforge (on Win 7, 64 Bit). The interesting part of the `.blg` when the compilation fails is

```
[394] bibtex.pm:1111> DEBUG - Parsing namestring 'al- Hakim, Tawfik'
[396] bibtex.pm:94> FATAL - Caught signal: SEGV
Likely your .bib has a very bad entry which causes libbtparse to crash: Invalid argument
```

I'm very sorry to provide only so little information and appreciate that occasional segmentation faults are really hard to investigate. Please let me know if you need any more information.

**Comment by plk** [comment]
Yes, unfortunately the btparse C library used does die on unusual input like this and all we can do is try to catch it and report that the .bib is likely bad. It is strange that it doesn't always fail though.

**Comment by moewew** [comment]
That's what made me wonder. If it were to fail at all time, one could say that this is an obviously wrongly formatted name entry (as a matter of fact, I think the input is wrong, but that is probably not the point), but four out of five times it seems to work, .. really puzzling.

**Comment by nickbart1980** [comment]
FWIW, I’m not seeing any of these issues with biber 2.2 / TeX Live 2015 on Mac OS X 10.10.5.

Also, I don’t feel the input is so unusual, let alone “wrong”. The space after a punctuation character is currently required as a workaround for what I’d call a bug in the parser. 

Note that there are also names that – even if that parser bug were fixed – have to be entered with a space between a particle ending with an apostrophe and the last name, e.g., “Lorenzo de’ Medici”.

@moewew: Can you reproduce the bug with `author = {de’ Medici, Lorenzo}`?

**Comment by moewew** [comment]
The segmentation fault did not appear with `author = {de’ Medici, Lorenzo}`, though of course I cannot say that the problem does not exists, only that I didn't observe any, given that the problematic entry works at times and throws an error at other times.

The result you are looking for is certainly nothing wrong or unusual, the problem is just that the current input is clearly (on my machine at least) parsed incorrectly as

```
{Hakim}{\bibinitperiod}{Tawfik}{\bibinitperiod}{al-}{\bibinitperiod}{}{}
```

i.e. the initials for last and first names are incorrectly reported as `bibinitperiod` in the `.bbl` instead of the desired

```
{Hakim}{H\bibinitperiod}{Tawfik}{T\bibinitperiod}{al-}{a\bibinitperiod}{}{}
```

**Comment by moewew** [comment]
I played around with this again since I remembered it when I read https://tex.stackexchange.com/q/496655/35864

The MWE still occasionally segfaults on Windows and always faulted on Ubuntu when I just tested.

I thought it might be useful to reproduce the underlying error with pure Perl `Text::BibTeX` but failed to do so
```perl
use Text::BibTeX::Name;

$name = Text::BibTeX::Name->new('al- Hakim, Tawfik');

@nameparts = qw(last first von jr);

foreach $namepart (@nameparts) {
  print "|" . $name->part ($namepart) . "|\n";
}
```
worked just fine.

Do you think it is possible to find out when we get the segfault? I know that it might be hard to fix the underlying issue here, but it would be nice to have the source of the issue documented.

**Comment by plk** [comment]
This looks like it's a problem with the default encoding which leads to invalid NFD UTF-8 decoding which eventually messes things up enough to cause this. Can you send me the `.bcf` of a failing MWE? I don't have things set up on Linux or Windows to run a full MWE.

**Comment by moewew** [comment]
I tested
```latex
\documentclass{article}
\usepackage[ngerman]{babel}
\usepackage[babel,german=guillemets]{csquotes} 
\usepackage[backend=biber, style=authoryear]{biblatex}

\usepackage{filecontents}
\begin{filecontents*}{\jobname.bib}
@book{hak,
  author = {al- Hakim, Tawfik},
  title = {Title},
  date = {2015},
}
\end{filecontents*}

\addbibresource{\jobname.bib}
\begin{document}
\cite{hak}
\printbibliography
\end{document}
```

- with `biblatex`+Biber `dev` on Windows is at https://gist.github.com/moewew/6c2c1415782a0fe09b09b4bb707c507a with occasional SEGV
- and release `biblatex`+Biber on Ubuntu https://gist.github.com/moewew/dfd396fa80d3c53c7b87934768e1e58a with consistent error.

Hopefully I'll be able to run a few more tests tomorrow with my self-compiled Linux binaries.

**Comment by plk** [comment]
Hmm, I can reproduce an error on OSX too, which is useful.

**Comment by plk** [comment]
I've put in a fix for this as it appears to be a specific problem with `btparse` and spaces after dash characters within name parts.

**Comment by moewew** [comment]
Did you manage to reproduce the error with pure `Text::BibTeX` methods? If so we could at least log the issue (even if a fix is tricky due to the implementation of the BibTeX parser in `btparse`).

Unfortunately, just removing spaces after dashes is too radical. In this case one would hope that thanks to the sapce `al-` would be parsed as the 'von part' (I'm not sure if that really must work), but there are other legitimate uses of `- ` as the following MWE shows
```latex
\documentclass{article}
\usepackage[ngerman]{babel}
\usepackage[babel,german=guillemets]{csquotes} 
\usepackage[backend=biber, style=authoryear]{biblatex}

\usepackage{filecontents}
\begin{filecontents*}{\jobname.bib}
@book{hak,
  author = {al- Hakim, Tawfik},
  title = {Title},
  date = {2015},
}
@book{bsp,
  author = {{Verein für Schäfer- und Hütehunde e.V.}},
  title  = {Title},
  date   = {2014},
}
\end{filecontents*}

\addbibresource{\jobname.bib}
\begin{document}
\cite{hak,bsp}
\printbibliography
\end{document}
```

**Comment by plk** [comment]
This is stranger than I thought - I also get intermittent success it seems. Looking into it.

**Comment by plk** [comment]
I think this is addressed now in DEV. There is certainly some intermittent thing going on inside `btparse` initials generation with certain malformed input. The malformed input was due to the default `noinit` regexps which were stripping `al-` even when it occurred as a prefix. This has been tuned and I no longer get any SEGVs.

**Comment by moewew** [comment]
Seems to work well now. Was this only because the `noinit` stripping would leave leading/trailing white spaces (or multiple spaces)? In that case it might be a nice idea to sanitise `$nd_namestr` again after stripping `noinit` in

https://github.com/plk/biber/blob/5840a04edf956ca3686f56861eb6aac2e30d21e1/lib/Biber/Input/file/bibtex.pm#L1515-L1602

**Comment by plk** [comment]
Yes, I think it will be a good idea to sanitise this after the `noinit` removal since these regexps can be arbitrary. This will be pushed shortly.

**Comment by plk** [comment]
Fixed in latest 2.13 release.
